pkg/kubectl/apply/strategy/retain_keys_test.go

```diff
@@ -0,0 +1,195 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxUnavailable: 1
+      maxSurge: 1
+`)
+			expected := create(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+    type: Recreate
+`)
+			run(strategy.Create(strategy.Options{}), recorded, local, remote, expected)
+		})
+	})
+
+	Context("where some fields are defined both locally and remotely", func() {
+		It("should merge those fields", func() {
+			recorded := create(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+`)
+			local := create(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxUnavailable: 2
+`)
+			remote := create(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxSurge: 1
+`)
+			expected := create(`
+apiVersion: extensions/v1beta1
+kind: Deployment
+spec:
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      maxUnavailable: 2
+      maxSurge: 1
+`)
+			run(strategy.Create(strategy.Options{}), recorded, local, remote, expected)
+		})
+	})
+
+	Context("where the elements are in a list and some fields are only defined remotely", func() {
+		It("should drop those fields ", func() {
+			recorded := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+`)
+			local := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        emptyDir:
+`)
+			remote := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        hostPath:
+          path: /tmp/cache-volume
+`)
+			expected := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        emptyDir:
+`)
+			run(strategy.Create(strategy.Options{}), recorded, local, remote, expected)
+		})
+	})
+
+	Context("where the elements are in a list", func() {
+		It("the fields defined both locally and remotely should be merged", func() {
+			recorded := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+`)
+			local := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        hostPath:
+          path: /tmp/cache-volume
+        emptyDir:
+`)
+			remote := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        hostPath:
+          path: /tmp/cache-volume
+          type: Directory
+`)
+			expected := create(`
+apiVersion: apps/v1beta1
+kind: Deployment
+spec:
+  template:
+    spec:
+      volumes:
+      - name: cache-volume
+        hostPath:
+          path: /tmp/cache-volume
+          type: Directory
+        emptyDir:
```

Comment:
Can we have both `hostPath` and `emptyDir`?

Reply:
No, this will fail validation on the server side, but the merge logic is correct here.

Reply:
Cool, that's what I expected, just wanted to be sure.